### PR TITLE
spki: use Result for Dyn*AlgorithmIdentifier traits

### PR DESCRIPTION
**Pull request by lumag**
There is a possibility that creating the AlgorithmIdentifier results in the Error. Change corresponding methods to return
Result<AlgorithmIdentifier>.

### CODE REVIEW

**Review by tarcieri — APPROVED** [review]
I'm fine with this. It would make the blanket impl more flexible, since we can use the `Encode` trait to build the `Any` value for `Params` (though that needs a bit more work).

@baloo are you okay with this?

File: spki/src/traits.rs [hunk 1]
```diff
@@ -123,11 +123,11 @@ where
     T: AssociatedAlgorithmIdentifier,
     T::Params: Into<Any>,
 {
-    fn algorithm_identifier(&self) -> AlgorithmIdentifierOwned {
-        AlgorithmIdentifierOwned {
+    fn algorithm_identifier(&self) -> Result<AlgorithmIdentifierOwned> {
+        Ok(AlgorithmIdentifierOwned {
             oid: T::ALGORITHM_IDENTIFIER.oid,
             parameters: T::ALGORITHM_IDENTIFIER.parameters.map(Into::into),
```
Comment (tarcieri) [inline spki/src/traits.rs:129]: With #976 landed you can get rid of the `T::Params: Into<Any>` bound and instead do:

```suggestion
            parameters: T::ALGORITHM_IDENTIFIER.parameters.as_ref().map(Any::encode_from).transpose()?,
```

File: spki/src/traits.rs [hunk 2]
```diff
@@ -159,10 +159,10 @@ where
     T: SignatureAlgorithmIdentifier,
     T::Params: Into<Any>,
 {
-    fn signature_algorithm_identifier(&self) -> AlgorithmIdentifierOwned {
-        AlgorithmIdentifierOwned {
+    fn signature_algorithm_identifier(&self) -> Result<AlgorithmIdentifierOwned> {
+        Ok(AlgorithmIdentifierOwned {
             oid: T::SIGNATURE_ALGORITHM_IDENTIFIER.oid,
             parameters: T::SIGNATURE_ALGORITHM_IDENTIFIER.parameters.map(Into::into),
```
Comment (tarcieri) [inline spki/src/traits.rs:165]: Same here:


```suggestion
            parameters: T::SIGNATURE_ALGORITHM_IDENTIFIER.parameters.as_ref().map(Any::encode_from),
```

File: spki/src/traits.rs (hunk 1 above)
Comment (lumag) [reply spki/src/traits.rs]: Implemented in a slightly different way.

File: spki/src/traits.rs (hunk 2 above)
Comment (lumag) [reply spki/src/traits.rs]: same

File: spki/src/traits.rs (hunk 1 above)
Comment (tarcieri) [reply spki/src/traits.rs]: That seems way more verbose, but I guess it gets the job done

Comment (lumag) [reply spki/src/traits.rs]: Maybe it's a matter of preference of simple `if let` compared to long chains of function calls.

Comment (tarcieri) [reply spki/src/traits.rs]: I imagine clippy will complain about it at some point in the future: https://github.com/rust-lang/rust-clippy/issues/4171

I'd consider the functional approach a bit more idiomatic.

Comment (lumag) [reply spki/src/traits.rs]: that lint talks about iterables. If you prefer, I can use your form, but IMHO it is less clear.

Comment (tarcieri) [reply spki/src/traits.rs]: If nothing else it's more consistent with the other uses of `transpose()` in this repo

Comment (lumag) [reply spki/src/traits.rs]: done

**Review by tarcieri — APPROVED** [review]
Still curious to hear from @baloo

**Comment by lumag** [comment]
> Still curious to hear from @baloo

Sure, me too.

**Comment by baloo** [comment]
Yeah that should work for me.

**Pull request merged**